src/toil/batchSystems/abstractGridEngineBatchSystem.py

```diff
@@ -275,6 +291,30 @@ def getJobExitCode(self, batchJobID):
             """
             raise NotImplementedError()
 
+        def getJobExitCodeWithRetries(self, batchJobID):
+            """
+            Returns job exit code. Implementation-specific; called by
+            AbstractGridEngineWorker.checkOnJobs()
+
+            :param string batchjobID: batch system job ID
+            """
+            exit_code = None
+            retries = 3
+            latest_err = None
+            while not exit_code and retries:
+                try:
+                    exit_code = self.getJobExitCode(batchJobID)
+                except subprocess.CalledProcessError as err:
+                    latest_err = err
+                    logger.error(
+                        "getJobExitCode failed with code %d: %s",
+                        err.returncode, err.output)
+                finally:
+                    retries -= 1
+            if latest_err:
```

Comment:
Same here, except with `and not exit_code==0`.

How would you feel about reducing the total code and trying to make one retry function for all of these cases?